### PR TITLE
Eliminate dependent-sum-template dependency

### PR DESCRIPTION
**Pull request by Bodigrim**
Cf. https://github.com/haskell/lsp/pull/377#issuecomment-998869748

Eliminating `dependent-sum-template`, `th-extras` and `th-abstraction` should make builds more robust and provide a better experience when upgrading to new GHCs.

### CODE REVIEW

**Review by jneira — APPROVED** [review]
wow, many many thanks, I had to chase maintainers of those packages several times to get support for newer ghcs
will wait for @michaelpj thoughts to merge

**Comment by michaelpj** [comment]
Nice! Using unsafeCoerce there wouldn't have occurred to me, but it seems entirely fine, and avoids the instance becoming huge. LGTM.

**Review by michaelpj — APPROVED** [review]

**Pull request merged**
